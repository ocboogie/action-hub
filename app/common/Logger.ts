type IListener = (msg: string) => void;

interface ILoggedError {
  category: string;
  msg: string;
}

export default class Logger {
  public history: ILoggedError[] = [];
  public historyLimit = 500;

  private categories: Map<string, IListener[]> = new Map();

  constructor(categories: string[]) {
    categories.forEach(category => {
      this.categories.set(category, []);
    });
  }

  public hasCategory(category: string): boolean {
    return this.categories.has(category);
  }

  public addListener(category: string, listener: IListener): () => void {
    const foundCategory = this.findCategory(category);
    foundCategory.push(listener);
    return () => {
      const index = foundCategory.indexOf(listener);
      if (index > -1) {
        foundCategory.splice(index, 1);
      }
    };
  }

  public report(category: string, msg: string) {
    this.history.unshift({ category, msg });
    if (this.history.length > this.historyLimit) {
      this.history = this.history.slice(0, this.historyLimit);
    }

    const foundCategory = this.findCategory(category);
    foundCategory.forEach(listener => {
      listener(msg);
    });
  }

  private findCategory(category: string): IListener[] {
    const foundCategory = this.categories.get(category);
    if (!foundCategory) {
      throw new TypeError(`No category named "${category}"`);
    }
    return foundCategory;
  }
}
