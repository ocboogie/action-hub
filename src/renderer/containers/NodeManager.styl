.NodeManager
    position: absolute
    top: 0
    bottom: 0
    left: 0
    right: 0
    overflow: hidden
    div
        height: 100%
        width: 100%