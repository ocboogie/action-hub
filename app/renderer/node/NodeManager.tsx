import CheerioReactBind from "cheerio-react-bind";
import * as React from "react";

import Logger from "../../common/Logger";
import { IRegistry } from "../registry";
import NodeTagRenderer from "./NodeTagRenderer";

export interface INodeManagerProps {
  $: CheerioStatic;
  registry: IRegistry;
  logger: Logger;
}

export default class NodeManager extends React.Component<INodeManagerProps> {
  constructor(props: INodeManagerProps) {
    super(props);
  }

  public errorHandler(msg: string) {
    this.props.logger.report("error", msg);
  }

  public render() {
    const errorHandler = this.errorHandler.bind(this);

    return (
      <>
        <CheerioReactBind
          $={this.props.$}
          $elem={this.props.$.root()
            .children()
            .first()}
          tagRendererArgs={{
            registry: this.props.registry
          }}
          tagRenderer={NodeTagRenderer}
          errorHandler={errorHandler}
        />
      </>
    );
  }
}
