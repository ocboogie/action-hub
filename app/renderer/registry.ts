import { INode } from "./node";

export interface IRegistry {
  nodes: {
    [nodeName: string]: INode;
  };
}
