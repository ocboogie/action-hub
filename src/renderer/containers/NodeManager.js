import React, { Component } from 'react';
import { connect } from 'react-redux';

import { nodeMap } from '../lib/node';
import NodeContainer from './NodeContainer';

class NodeManager extends Component {
    render() {
        return (
            <div>
                <NodeContainer node={this.props.node} />
            </div>
        );
    }
}

const mapStateToProps = (state) => (state)

export default connect(mapStateToProps)(NodeManager)