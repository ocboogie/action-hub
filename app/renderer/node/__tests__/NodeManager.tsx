import * as Cheerio from "cheerio";
// tslint:disable-next-line:no-implicit-dependencies
import { mount } from "enzyme";
import * as React from "react";

import Logger from "../../../common/Logger";
import { IRegistry } from "../../registry";
import NodeManager from "../NodeManager";
import childrenNode from "./resources/mockNodes/children";
import textNode from "./resources/mockNodes/text";

const registry: IRegistry = {
  nodes: {
    text: textNode,
    children: childrenNode
  }
};

const logger = new Logger(["error"]);

describe("rendering", () => {
  test("renders correct nodes", () => {
    const $mock = Cheerio.load(
      `
      <text text="foo" />
  `,
      { xmlMode: true }
    );
    const wrapper = mount(
      <NodeManager $={$mock} registry={registry} logger={logger} />
    );

    expect(wrapper.contains(<h1>foo</h1>)).toBeTruthy();
  });

  test("updates", () => {
    const $mock = Cheerio.load(
      `
    <children>
    </children>
  `,
      { xmlMode: true }
    );
    const wrapper = mount(
      <NodeManager $={$mock} registry={registry} logger={logger} />
    );

    $mock("children").append('<text text="foo" />');
    $mock("children").update();
    expect(wrapper.html()).toMatchSnapshot();
  });
});

describe("error handling", () => {
  test("logs when invalid args are supplied to a node", () => {
    (Error.prototype as any).suppressReactErrorLogging = true;
    // Related issue https://github.com/airbnb/enzyme/issues/1255
    const $mock = Cheerio.load(
      `
        <text />
    `,
      { xmlMode: true }
    );
    const mockLogger = { report: jest.fn() };
    mount(
      <NodeManager $={$mock} registry={registry} logger={mockLogger as any} />
    );

    expect(mockLogger.report.mock.calls[0]).toEqual([
      "error",
      'There was an error rendering a tag at "/": "Invalid arguments: child "text" fails because ["text" is required]."'
    ]);
    (Error.prototype as any).suppressReactErrorLogging = false;
  });

  test("logs when unknown node name is found", () => {
    (Error.prototype as any).suppressReactErrorLogging = true;
    // Related issue https://github.com/airbnb/enzyme/issues/1255
    const $mock = Cheerio.load(
      `
        <notARealTag />
    `,
      { xmlMode: true }
    );
    const mockLogger = { report: jest.fn() };
    mount(
      <NodeManager $={$mock} registry={registry} logger={mockLogger as any} />
    );

    expect(mockLogger.report.mock.calls[0]).toEqual([
      "error",
      'There was an error rendering a tag at "/": "Unknown node name "notarealtag"."'
    ]);
    (Error.prototype as any).suppressReactErrorLogging = false;
  });
});
