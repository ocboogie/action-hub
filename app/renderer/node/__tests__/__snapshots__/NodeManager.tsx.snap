// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering updates 1`] = `"<div>start<div><h1>foo</h1></div>end</div>"`;
